_imagingft.c

```diff
@@ -188,60 +205,284 @@ font_getchar(PyObject* string, int index, FT_ULong* char_out)
+        (*glyph_info)[i].x_offset = glyphs[i].x_offset;
+        (*glyph_info)[i].x_advance = glyphs[i].x_advance;
+        (*glyph_info)[i].y_offset = glyphs[i].y_offset;
+        (*glyph_info)[i].cluster = glyphs[i].cluster;
+    }
+
+failed:
+    raqm_destroy (rq);
+    return count;
+}
+#endif
+
+static size_t
+text_layout_fallback(PyObject* string, FontObject* self, const char* dir,
+            PyObject *features ,GlyphInfo **glyph_info, int mask)
+{
+    int error, load_flags;
     FT_ULong ch;
-    FT_Face face;
-    int xoffset, yoffset;
+    Py_ssize_t count;
+    FT_GlyphSlot glyph;
     FT_Bool kerning = FT_HAS_KERNING(self->face);
     FT_UInt last_index = 0;
+    int i;
 
-    /* calculate size and bearing for a given string */
-
-    PyObject* string;
-    if (!PyArg_ParseTuple(args, "O:getsize", &string))
-        return NULL;
-
+    if (features != Py_None || dir != NULL) {
+      PyErr_SetString(PyExc_KeyError, "setting text direction or font features is not supported without libraqm");
+    }
 #if PY_VERSION_HEX >= 0x03000000
     if (!PyUnicode_Check(string)) {
 #else
     if (!PyUnicode_Check(string) && !PyString_Check(string)) {
 #endif
         PyErr_SetString(PyExc_TypeError, "expected string");
-        return NULL;
+        return 0;
     }
 
+    count = 0;
+    while (font_getchar(string, count, &ch)) {
+       count++;
+    }
+    if (count == 0) {
+        return 0;
+    }
+
+    (*glyph_info) = PyMem_New(GlyphInfo, count);
+    if ((*glyph_info) == NULL) {
+        PyErr_SetString(PyExc_MemoryError, "PyMem_New() failed");
+        return 0;
+    }
+
+    load_flags = FT_LOAD_RENDER|FT_LOAD_NO_BITMAP;
+    if (mask) {
+        load_flags |= FT_LOAD_TARGET_MONO;
+    }
+    for (i = 0; font_getchar(string, i, &ch); i++) {
+        (*glyph_info)[i].index = FT_Get_Char_Index(self->face, ch);
+        error = FT_Load_Glyph(self->face, (*glyph_info)[i].index, load_flags);
+        if (error) {
+            geterror(error);
+            return 0;
+        }
+        glyph = self->face->glyph;
+        (*glyph_info)[i].x_offset=0;
+        (*glyph_info)[i].y_offset=0;
+        if (kerning && last_index && (*glyph_info)[i].index) {
+            FT_Vector delta;
+            if (FT_Get_Kerning(self->face, last_index, (*glyph_info)[i].index,
+                           ft_kerning_default,&delta) == 0)
+            (*glyph_info)[i-1].x_advance += PIXEL(delta.x);
+        }
+
+        (*glyph_info)[i].x_advance = glyph->metrics.horiAdvance;
+        last_index = (*glyph_info)[i].index;
+        (*glyph_info)[i].cluster = ch;
+    }
+    return count;
+}
+
+static size_t
+text_layout(PyObject* string, FontObject* self, const char* dir,
+            PyObject *features, GlyphInfo **glyph_info, int mask)
+{
+    size_t count;
+#ifdef HAVE_RAQM
+    count = text_layout_raqm(string, self, dir, features, glyph_info,  mask);
+#else
+    count = text_layout_fallback(string, self, dir, features, glyph_info, mask);
```

Comment:
Say an existing user runs `pip install --upgrade pillow` without installing raqm. Is the new text rendering code they will be running tested?